python/ray/workflow/step_executor.py

```diff
@@ -195,16 +197,41 @@ async def _write_step_inputs(wf_storage: workflow_storage.WorkflowStorage,
 
 
 def commit_step(store: workflow_storage.WorkflowStorage, step_id: "StepID",
-                ret: Union["Workflow", Any], *,
-                exception: Optional[Exception]):
+                ret: Union["Workflow", Any], *, exception: Optional[Exception],
+                checkpoint: "CheckpointModeType"):
     """Checkpoint the step output.
     Args:
         store: The storage the current workflow is using.
         step_id: The ID of the step.
         ret: The returned object of the workflow step.
         exception: The exception caught by the step.
+        checkpoint: The checkpoint mode.
     """
     from ray.workflow.common import Workflow
+
+    if checkpoint == CheckpointMode.DEFAULT.value:
+        # Default value implies "SYNC".
+        checkpoint = CheckpointMode.SYNC.value
+    if checkpoint == CheckpointMode.SKIP.value:
+        # Check if the checkpoint options are valid.
+        if isinstance(ret, Workflow):
+            for w in ret._iter_workflows_in_dag():
+                print(w.data.step_options.checkpoint)
+                if w != ret and w.data.step_options.checkpoint:
+                    raise ValueError(
+                        f"Step '{w.step_id}' requires checkpointing, however "
+                        f"this is invalid. It is because the parent step "
+                        f"'{step_id}' skips checkpointing, and the DAG of "
+                        f"the sub-workflow is not preserved in the storage. "
+                        f"Without the DAG, we cannot cover '{w.step_id}' "
+                        f"from its checkpoint. NOTE: This rule does not "
+                        f"affect the output step of the sub-workflow, "
+                        f"because once the output step finished, we can "
+                        f"recover from the checkpoint of the output step "
+                        f"instead of the parent step.")
```

Comment:
```suggestion
                        f"Step '{w.step_id}' requires checkpointing, however "
                        f"this is invalid. This is because the parent step "
                        f"'{step_id}' skips checkpointing, so the DAG of "
                        f"the returned sub-workflow is not preserved in storage. "
                        f"Without the DAG, we cannot cover '{w.step_id}' "
                        f"from its checkpoint. NOTE: This rule does not "
                        f"affect the output step of the sub-workflow, "
                        f"because once the output step finished, we can "
                        f"recover from the checkpoint of the output step "
                        f"instead of the parent step.")
```

Hmm just so I understand correctly, I thought that actually this is meaningless, even though `child` is the output step:

```
parent(checkpoint=False):
  return child(checkpoint=True)
```

But here it does work because `child` is the output of the grandparent:
```
grandparent(checkpoint=True):
  return parent(checkpoint=False):

parent(checkpoint=False):
  return child(checkpoint=True)
```

Is that correct?

Reply:
I think the description is not very clear here, and I just find there is actually a flaw in the implementation. Let me fix it.